mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageIdMapper.java

```diff
@@ -0,0 +1,279 @@
+            .concatMap(mailboxId -> flagsUpdateWithRetry(newState, updateMode, mailboxId, messageId))
+            .collect(ImmutableListMultimap.toImmutableListMultimap(Pair::getLeft, Pair::getRight));
+    }
+
+    private Flux<Pair<MailboxId, UpdatedFlags>> flagsUpdateWithRetry(Flags newState, MessageManager.FlagsUpdateMode updateMode, MailboxId mailboxId, MessageId messageId) {
+        return updateFlags(mailboxId, messageId, newState, updateMode)
+            .retry(NUM_RETRIES)
+            .onErrorResume(MailboxDeleteDuringUpdateException.class, e -> {
+                LOGGER.info("Mailbox {} was deleted during flag update", mailboxId);
+                return Mono.empty();
+            })
+            .flatMapIterable(Function.identity())
+            .map(pair -> buildUpdatedFlags(pair.getRight(), pair.getLeft()));
+    }
+
+    private Pair<MailboxId, UpdatedFlags> buildUpdatedFlags(ComposedMessageIdWithMetaData composedMessageIdWithMetaData, Flags oldFlags) {
+        return Pair.of(composedMessageIdWithMetaData.getComposedMessageId().getMailboxId(),
+            UpdatedFlags.builder()
+                .uid(composedMessageIdWithMetaData.getComposedMessageId().getUid())
+                .messageId(composedMessageIdWithMetaData.getComposedMessageId().getMessageId())
+                .modSeq(composedMessageIdWithMetaData.getModSeq())
+                .oldFlags(oldFlags)
+                .newFlags(composedMessageIdWithMetaData.getFlags())
+                .build());
+    }
+
+    private Mono<List<Pair<Flags, ComposedMessageIdWithMetaData>>> updateFlags(MailboxId mailboxId, MessageId messageId, Flags newState, MessageManager.FlagsUpdateMode updateMode) {
+        PostgresMailboxId postgresMailboxId = (PostgresMailboxId) mailboxId;
+        PostgresMessageId postgresMessageId = (PostgresMessageId) messageId;
+        return mailboxMessageDAO.findMetadataByMessageId(postgresMessageId, postgresMailboxId)
+            .flatMap(oldComposedId -> updateFlags(newState, updateMode, postgresMailboxId, oldComposedId), ReactorUtils.DEFAULT_CONCURRENCY)
+            .switchIfEmpty(Mono.error(MailboxDeleteDuringUpdateException::new))
+            .collectList();
+    }
+
+    private Mono<Pair<Flags, ComposedMessageIdWithMetaData>> updateFlags(Flags newState, MessageManager.FlagsUpdateMode updateMode, PostgresMailboxId mailboxId, ComposedMessageIdWithMetaData oldComposedId) {
+        FlagsUpdateCalculator flagsUpdateCalculator = new FlagsUpdateCalculator(newState, updateMode);
+        Flags newFlags = flagsUpdateCalculator.buildNewFlags(oldComposedId.getFlags());
+        if (identicalFlags(oldComposedId, newFlags)) {
+            return Mono.just(Pair.of(oldComposedId.getFlags(), oldComposedId));
+        } else {
+            return modSeqProvider.nextModSeqReactive(mailboxId)
+                .flatMap(newModSeq -> {
+                    // TODO bug?
+                    // ModSeq previousModseq = oldComposedId.getModSeq();
```

Comment:
what do you mean here @chibenwa?

Reply:
If this works you can remove the comment IMO